### PR TITLE
feat(editor): signature help

### PR DESCRIPTION
**Pull request by zbaylin**
⚠️ Again, very WIP ⚠️ 

This creates a `Feature_SignatureHelp` module that will be the integration of the information we get from the extension host

### CODE REVIEW

**Comment by zbaylin** [comment]
I think I'm getting a race condition bug where, when the trigger key is hit, the buffer isn't updated in the extension host so the request returns `null`. @bryphe have you dealt with any similar bugs in the past that might help me figure this one out?

File: src/Core/Json.re [hunk 1]
```diff
@@ -18,10 +18,12 @@ module Decode = {
         string => result(value, string)
     );
 
-    let of_file = file =>
+    let of_file = file => {
+      Console.log(file);
```
Comment (bryphe) [inline src/Core/Json.re:22]: We'll need to remove the debug logging

File: src/Exthost/SignatureHelp.re [hunk 2]
```diff
@@ -65,15 +65,30 @@ module RequestContext = {
 module ParameterInformation = {
   [@deriving show]
   type t = {
-    label: string,
+    label: [ | `String(string) | `Range(int, int)],
     documentation: option(MarkdownString.t),
   };
 
+  let label =
```
Comment (bryphe) [inline src/Exthost/SignatureHelp.re:72]: Thanks for fixing this up!

File: src/Feature/Hover/Feature_Hover.re [hunk 3]
```diff
@@ -34,15 +34,8 @@ let initial = {
...
-  let current = ref(0);
-
-  let get = () => {
-    let id = current^;
-    current := id + 1;
-    id;
-  };
-};
+module IDGenerator =
+  Oni_Core.Utility.IDGenerator.Make({});
```
Comment (bryphe) [inline src/Feature/Hover/Feature_Hover.re:38]: We actually have a similar concept in Revery, too (`Revery.UniqueId.Make()`)

File: src/Components/HoverView.re [hunk 4]
```diff
@@ -0,0 +1,152 @@
+open Revery;
```
Comment (bryphe) [inline src/Components/HoverView.re:1]: Nice! Thanks for refactoring this out to a common place - good chunk of code to share between hover and signature help 👍

File: src/Feature/Hover/Feature_Hover.re (hunk 3 above)
Comment (zbaylin) [reply src/Feature/Hover/Feature_Hover.re:38]: Oh cool, didn't notice that!

File: src/Feature/SignatureHelp/Feature_SignatureHelp.rei [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  | InfoReceived({
+      signatures: list(Exthost.SignatureHelp.Signature.t),
+      activeSignature: int,
+      activeParameter: int,
+      requestID: int,
+    })
+  | EmptyInfoReceived(int)
+  | RequestFailed(string)
+  | SignatureIncremented
+  | SignatureDecremented;
```
Comment (bryphe) [inline src/Feature/SignatureHelp/Feature_SignatureHelp.rei:34]: Might be worth renaming this such that they describe the event that occurs - like `SignatureIncrementClicked`/`SignatureDecrementClicked`. That way, if there ends up being some subtle difference in behavior between clicking and the command - it's obvious where to handle it.

File: src/Model/ContextKeys.re [hunk 6]
```diff
@@ -25,48 +25,53 @@ let menus =
...
+        }
+      ),
+      bool("visualMode", state =>
+        switch (ModeManager.current(state)) {
+        | TerminalVisual
+        | Visual => true
+        | _ => false
+        }
+      ),
+      bool("parameterHintsVisible", state =>
```
Comment (bryphe) [inline src/Model/ContextKeys.re:70]: Thanks for adding this key!

File: src/Service/Exthost/Service_Exthost.re [hunk 7]
```diff
@@ -108,6 +108,27 @@ module Effects = {
         );
       });
     };
+
+    let provideSignatureHelp =
```
Comment (bryphe) [inline src/Service/Exthost/Service_Exthost.re:112]: Nice work wiring this up! 🎉

File: src/Store/KeyBindingsStoreConnector.re [hunk 8]
```diff
@@ -459,6 +459,18 @@ let start = maybeKeyBindingsFilePath => {
         command: Feature_Layout.Commands.toggleMaximize.id,
         condition: windowCommandCondition,
       },
+      {
+        key: "<A-DOWN>",
+        command: Feature_SignatureHelp.Commands.incrementSignature.id,
+        condition:
+          "editorTextFocus && parameterHintsVisible" |> WhenExpr.parse,
+      },
```
Comment (bryphe) [inline src/Store/KeyBindingsStoreConnector.re:467]: Thanks for adding these!

File: src/Core/Json.re (hunk 1 above)
Comment (zbaylin) [reply src/Core/Json.re]: Oops! I thought I had removed all of them. Thanks for pointing this one out! Fixed now 👍

File: src/Feature/SignatureHelp/Feature_SignatureHelp.rei (hunk 5 above)
Comment (zbaylin) [reply src/Feature/SignatureHelp/Feature_SignatureHelp.rei]: Ah you're right. I changed it in the latest commit. Thanks!

**Pull request merged**